plugins/SpectrumAnalyzer/SaProcessor.cpp

```diff
@@ -79,169 +87,234 @@ SaProcessor::~SaProcessor()
+									accR += (band_end - band_start) * m_normSpectrumR[i];
+								}
+								else
+								{
+									// Band ends in the next pixel -- finalize the current pixel.
+									// Make sure contribution is split correctly on pixel boundary.
+									accL += ((int)band_end - band_start) * m_normSpectrumL[i];
+									accR += ((int)band_end - band_start) * m_normSpectrumR[i];
 
-								// save remaining portion of the band for the following band / pixel
-								accL = (band_end - (int)band_end) * m_normSpectrumL[i];
-								accR = (band_end - (int)band_end) * m_normSpectrumR[i];
+									if (target >= 0 && target < waterfallWidth()) {pixel[target] = makePixel(accL, accR);}
+
+									// save remaining portion of the band for the following band / pixel
+									accL = (band_end - (int)band_end) * m_normSpectrumL[i];
+									accR = (band_end - (int)band_end) * m_normSpectrumR[i];
+								}
 							}
 						}
-					}
-					else
-					{
-						// Linear: always draws one or more pixels per band
-						for (target = (int)band_start; target < band_end; target++)
+						else
 						{
-							if (target >= 0 && target < binCount())
+							// Linear: always draws one or more pixels per band
+							for (target = (int)band_start; target < band_end; target++)
 							{
-								pixel[target] = makePixel(m_normSpectrumL[i], m_normSpectrumR[i]);
+								if (target >= 0 && target < waterfallWidth())
+								{
+									pixel[target] = makePixel(m_normSpectrumL[i], m_normSpectrumR[i]);
+								}
 							}
 						}
 					}
+
+					// Copy work buffer to result buffer. Done only if requested, so
+					// that time isn't waster on updating faster than display FPS.
+					// (The copy is about as expensive as the movement.)
+					if (m_flipRequest)
+					{
+						m_history = m_history_work;
+						m_flipRequest = false;
+					}
 				}
-			}
-			#ifdef SA_DEBUG
-				// report FFT processing speed
-				start_time = std::chrono::high_resolution_clock::now().time_since_epoch().count() - start_time;
-				std::cout << "Processed " << m_framesFilledUp << " samples in " << start_time / 1000000.0 << " ms" << std::endl;
-			#endif
-
-			// clean up before checking for more data from input buffer
-			m_framesFilledUp = 0;
-		}
-	}
+				// clean up before checking for more data from input buffer
+				const unsigned int overlaps = m_controls->m_windowOverlapModel.value();
+				if (overlaps == 1)	// Discard buffer, each sample used only once
+				{
+					m_framesFilledUp = 0;
+				}
+				else
+				{
+					// Drop only a part of the buffer from the beginning, so that new
+					// data can be added to the end. This means the older samples will
+					// be analyzed again, but in a different position in the window,
+					// making short transient signals show up better in the waterfall.
+					const unsigned int drop = m_inBlockSize / overlaps;
+					m_bufferL.erase(m_bufferL.begin(), m_bufferL.begin() + drop);
```

Comment:
Ah, I somehow missed the point that `erase()` eats elements from the start of the array, so it will be reallocated, causing another reallocation on `resize()`.
`std::move` looks good and also "self-documenting", I just did not know it even existed. :)

Reply:
What if `m_framesFilledUp` is less than `drop`? In that case, this code will drop samples without changing `m_framesFilledUp`.

Reply:
Thad should never happen: the cleanup / buffer rotation is done after analysis, and analysis is done only after `m_framesFilledUp` reaches the input block size. The `drop` itself depends on the input block size so it cannot become larger.

Reply:
If so, isn't the `if` statement in `if (drop <= m_framesFilledUp) {m_framesFilledUp -= drop;}` redundant?

Reply:
Hmm, that's true. I remember I ran into a problem when `m_framesFilledUp` became negative, so the check is mainly for that. I can't remember the exact reason, though, maybe the original issue was meanwhile fixed.. :thinking:

Reply:
After the change to `std::move` I removed the extra condition and did not see any problems, so I'm keeping the simplified version.